### ISSUE TITLE
Error propagation operator support

### ISSUE THREAD

**Issue by mdesantis** [djc/askama#431]
Hello!

I have a template similar to this one (edited to go to the point):

```
{{ request.url_for("asd", ["dsa"])? }}
```

Which generates a syntax error on Askama:

```
error: proc-macro derive panicked
 --> src/templates/home.rs:5:10
  |
5 | #[derive(AskamaActixTemplate)]
  |          ^^^^^^^^^^^^^^^^^^^
  |
  = help: message: unable to parse template:
          
          "{{ request.url_for(\"asd\", [\"dsa\"])? }}"

error[E0599]: no method named `into_response` found for struct `templates::home::Template<'_>` in the current scope
  --> src/responders/home.rs:11:6
   |
11 |     .into_response()
   |      ^^^^^^^^^^^^^ method not found in `templates::home::Template<'_>`
   | 
  ::: src/templates/home.rs:7:1
   |
7  | pub struct Template<'a> {
   | -----------------------
   | |
   | method `into_response` not found for this
   | doesn't satisfy `_: TemplateIntoResponse`
   | doesn't satisfy `_: askama_actix::Template`
   |
   = note: the method `into_response` exists but the following trait bounds were not satisfied:
           `templates::home::Template<'_>: askama_actix::Template`
           which is required by `templates::home::Template<'_>: TemplateIntoResponse`
```

Shouldn't the error propagation operator (`?`) be supported? Is there any workaround for it? I'm new to Rust so I'm pretty confused :D

**Comment by djc** [comment]
The template language does not support all of the Rust language -- it doesn't actually support the `?` operator. We'll have to think about whether adding support for it makes sense.

**Comment by vallentin** [comment]
What would the expected outcome be? That it propagates all the way up and returns an `Err` for [`render()`]? If yes, then there additionally, needs to be a way to specify the `Err` type returned by [`render()`], which would also require `impl From<`[`Error`]`> for CustomError`.

[`render()`]: https://docs.rs/askama/0.10.5/askama/trait.Template.html#method.render
[`Error`]: https://docs.rs/askama/0.10.5/askama/enum.Error.html

Potential candidates could be:

- `template(err = CustomError)`
- `impl MyTemplate { type Err = CustomError; }`

**Comment by djc** [comment]
Yeah, I suppose we'd add an associated error type to `Template`, which the `Error` type gets parametrized with. It seems pretty messy.

**Comment by vallentin** [comment]
We could make it optional. So if an `err = CustomError` or in the future an associated type is used, then `render()` uses that type. If it is not specified, then it remains as is, and returns [`askama::Error`].

[`askama::Error`]: https://docs.rs/askama/0.10.5/askama/enum.Error.html

**Comment by blazzy** [comment]
I take this to mean there is no way to surface an error on render right now whether an operator exists or not.

I had been getting away with passing references to my application's internal data structures to templates. But it went awry once I introduced functions that could fail. For example:

```jinja
{%~ for operation in api.operations %}
  /// {{operation.method}} {{operation.path-}}

  {% for param in operation.params%}
     {{param.name}}: {{param.fn_that_can_error()}}
  {%~ endfor -%}
{% endfor -%}
```

Given the desire to abort when `fn_that_can_error` has an error, I find myself creating parallel data structures that invoke `fn_that_can_error` and resolve it to to strings ahead of time. It's inconvenient, but maybe I should have been using some kind of intermediate data transfer object anyway?

**Comment by djc** [comment]
I do feel like getting fallibility out of the way before templating generally makes sense? Doesn't mean you have to resolve everything to strings, just that, if you want to render a state, conceptually it makes sense to me to "materialize" that state first so that template rendering doesn't need to encompass your whole processing pipeline.
